### ISSUE TITLE
Getting started with opentelemetry-spark

### ISSUE THREAD

**Issue by progxaker** [keytiong/opentelemetry-spark#2]
Hello @keytiong. First of all, thank you for all your time and effort in keeping this repository up to date. I appreciate it as this is the only repository that was found for the query `onTaskStart io.opentelemetry org.apache.spark` on GitHub :slightly_smiling_face: and it *seems* (this is the main topic...) that it implements real distributed tracing in Spark.

---

I'm trying to run opentelemetry-spark on **Spark 3.4.1** (based on your commit bbe2095bc3914f55c52df525485f63fc595c6ea8, it's supported), but running a bunch of problems.
1. If I run the command described in `README.md`: `./gradlew build`,
    the result is:

       Error: Could not find or load main class org.gradle.wrapper.GradleWrapperMain
       Caused by: java.lang.ClassNotFoundException: org.gradle.wrapper.GradleWrapperMain

    Workaround: run locally installed **Gradle 8.2.1**.

2. Command: `gradle --no-daemon clean build`
    Result:

       To honour the JVM settings for this build a single-use Daemon process will be forked. For more on this, please refer to https://docs.gradle.org/8.2.1/userguide/gradle_daemon.html#sec:disabling_the_daemon in the Gradle documentation.
       Daemon will be stopped at the end of the build
        ﻿ 
       FAILURE: Build failed with an exception.
        ﻿ 
       * What went wrong:
       Could not determine the dependencies of task ':spotlessCheck'.
       > Could not create task ':spotless JavaCheck'.
           > Could not create task ':spotlessJava'.
               > java.nio.file.NoSuchFileException: /buildscripts/spotless.license.java
 
       * Try:
       > Run with --stacktrace option to get the stack trace.
       > Run with --info or --debug option to get more log output.
       > Run with --scan to get full insights.
       > Get more help at https://help.gradle.org.

       Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

       You can use '-warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

       For more on this, please refer to https://docs.gradle.org/8.2.1/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

       BUILD FAILED in 10s

    Workaround: comment out the line below. https://github.com/keytiong/opentelemetry-spark/blob/ec350164e4683010b6b83950f9465f6e24fb5d15/build.gradle#L59

3. I built the library, copied it into an image that includes Spark, and used the following parameters as specified in the README:

       -javaagent:/opt/spark/jars/opentelemetry-javaagent.jar
       -Dotel.javaagent.extensions=/opt/spark/jars/opentelemetry-spark-0.10.0-SNAPSHOT-all.jar
       -Dotel.java.global-autoconfigure.enabled=true

    but there are no related spans (I first tried with [Azure Monitor OT Distro](https://learn.microsoft.com/en-us/azure/azure-monitor/app/opentelemetry-enable?tabs=java#what-is-the-azure-monitor-opentelemetry-distro), then with the provided OpenTelemetry library in `build/libs/`, which required me to upgrade **Java** from **11** to **17**) such as <code><a href="https://github.com/keytiong/opentelemetry-spark/blob/b9ebe2fc9851c0f05b4f0b5bff82c8e387c4f359/src/main/java/io/opentelemetry/javaagent/instrumentation/spark/v3_1/SparkEventListener.java#L41">spark_job</a></code> or <code><a href="https://github.com/keytiong/opentelemetry-spark/blob/b9ebe2fc9851c0f05b4f0b5bff82c8e387c4f359/src/main/java/io/opentelemetry/javaagent/instrumentation/spark/v3_1/SparkEventListener.java#L41">spark_stage</a></code>, either in a tracing system or with `-Dotel.traces.exporter=logging`. Other traces, such as HTTP requests from Spark to Azure Storage, are there.

I'm not sure what I'm doing wrong... I thought this problem was in the tricky App Insights (they like rename [standard properties](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/315d21158d9c06fa8c941379ef48127d630effac/instrumentation/methods/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/methods/MethodInstrumentationModule.java#L24) to [their own](https://github.com/microsoft/ApplicationInsights-Java/blob/2c0559fad9f6a0d83e5792c80181fc77aced95b7/agent/instrumentation/methods/src/main/java/io/opentelemetry/javaagent/instrumentation/methods/ai/MethodInstrumentationModule.java#L28)), but with the community OpenTelemetry agent it's the same problem.
If you wouldn't mind helping me with this or providing additional steps on how to debug/deploy your code, I'd really appreciate it.

Thanks in advance :slightly_smiling_face: 

Regards,
Eduard

**Comment by progxaker** [comment]
Command: `gradle clean test`.
Result:
```
﻿Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.
﻿
You can use use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/8.2.1/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 7s
5 actionable tasks: 5 executed
```

**Comment by keytiong** [comment]
@progxaker. Thanks for reporting the build issue. I'm new to Gradle and didn't realise that gradle-wrapper.jar and gradle-wrapper.properties are meant to be committed to to VCS in order for gradlew to work. I also fixed the spotless license file reported. I've setup a GitHub action to build the binary going forward instead of building on my local machine.

**Comment by keytiong** [comment]
@progxaker 

Are you running with the open source Apache Spark 3.4.1 or another Spark 3.4.1 distribution?

The code has been developed with the following configuration

- Apache Spark 3.4.1
- Apache YARN Hadoop 3.3.6
- OpenTelemetry javaagent 1.31.0
- Tested with `org.apache.spark.examples.SparkPi` in cluster mode
- Single node pseudo cluster on local machine

You may want to add the following system property to the `spark.driver.extraJavaOptions` and `spark.executor.extraJavaOptons`.

`-Dotel.javaagent.debug=true`

That will emit verbose OpenTelemetry javaagent debug messages to system.err which you can view in YARN Resource Manager console. Hopefully that may shade some light on any exception or error during instrumentation

**Comment by progxaker** [comment]
Hi @keytiong.

> I also fixed the spotless license file reported.

Thank you)

> Are you running with the open source Apache Spark 3.4.1 or another Spark 3.4.1 distribution?

I'm using the open-source one ([spark-3.4.1-bin-hadoop3.tgz](https://archive.apache.org/dist/spark/spark-3.4.1/)) on a Kubernetes cluster.

> -Dotel.javaagent.debug=true

Hm, I'll try that and let you know.

Thanks for the response :slightly_smiling_face:

**Comment by progxaker** [comment]
Thanks @keytiong, with the DEBUG log level I found three stack dumps (all with `NullPointerException`):

<details>
<summary>onJobStart</summary>

```
[otel.javaagent 2024-02-13 19:25:24:094 +0000] [dag-scheduler-event-loop] DEBUG io.opentelemetry.javaagent.bootstrap.ExceptionLogger - Failed to handle exception in instrumentation for org.apache.spark.scheduler.LiveListenerBus
java.lang.NullPointerException
        at io.opentelemetry.javaagent.shaded.instrumentation.api.internal.cache.weaklockfree.AbstractWeakConcurrentMap.put(AbstractWeakConcurrentMap.java:146)
        at io.opentelemetry.javaagent.shaded.instrumentation.api.internal.cache.weaklockfree.WeakConcurrentMap.put(WeakConcurrentMap.java:48)
        at io.opentelemetry.javaagent.shaded.instrumentation.api.internal.cache.weaklockfree.WeakConcurrentMap$WithInlinedExpunction.put(WeakConcurrentMap.java:206)
        at io.opentelemetry.javaagent.shaded.instrumentation.api.internal.cache.WeakLockFreeCache.put(WeakLockFreeCache.java:31)
        at io.opentelemetry.javaagent.bootstrap.field.VirtualFieldImpl$org$apache$spark$scheduler$ActiveJob$io$opentelemetry$javaagent$shaded$io$opentelemetry$context$Context.mapPut(VirtualFieldImplementationsGenerator.java:304)
        at io.opentelemetry.javaagent.bootstrap.field.VirtualFieldImpl$org$apache$spark$scheduler$ActiveJob$io$opentelemetry$javaagent$shaded$io$opentelemetry$context$Context.realPut(VirtualFieldImplementationsGenerator.java)
        at io.opentelemetry.javaagent.bootstrap.field.VirtualFieldImpl$org$apache$spark$scheduler$ActiveJob$io$opentelemetry$javaagent$shaded$io$opentelemetry$context$Context.set(VirtualFieldImplementationsGenerator.java:284)
        at io.opentelemetry.javaagent.instrumentation.spark.ApacheSparkSingletons.setJobContext(ApacheSparkSingletons.java:117)
        at io.opentelemetry.javaagent.instrumentation.spark.v3_1.SparkEventListener.onJobStart(SparkEventListener.java:71)
        at io.opentelemetry.javaagent.instrumentation.spark.v3_1.SparkEventListener.handleSparkListenerEvent(SparkEventListener.java:254)
        at org.apache.spark.scheduler.LiveListenerBus.post(LiveListenerBus.scala:126)
        at org.apache.spark.scheduler.DAGScheduler.handleJobSubmitted(DAGScheduler.scala:1331)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2991)
        at org.apache.spark.scheduler.DAGSchedulerEventProcess Loop.onReceive(DAGScheduler.scala:2982)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2971)
        at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
```

</details>

<details>
<summary>onStageSubmitted</summary>

```
[otel.javaagent 2024-02-13 19:25:24:105 +0000] [dag-scheduler-event-loop] DEBUG io.opentelemetry.javaagent.bootstrap.ExceptionLogger - Failed to handle exception in instrumentation for org.apache.spark.scheduler.LiveListenerBus
java.lang.NullPointerException
        at io.opentelemetry.javaagent.shaded.instrumentation.api.internal.cache.weaklockfree.AbstractWeakConcurrentMap.getIfPresent(AbstractWeakConcurrentMap.java:113)
        at io.opentelemetry.javaagent.shaded.instrumentation.api.internal.cache.weaklockfree.WeakConcurrentMap.getIfPresent(WeakConcurrentMap.java:48)
        at io.opentelemetry.javaagent.shaded.instrumentation.api.internal.cache.weaklockfree.WeakConcurrentMap$WithInlinedExpunction.getIfPresent(WeakConcurrentMap.java:194)
        at io.opentelemetry.javaagent.shaded.instrumentation.api.internal.cache.WeakLockFreeCache.get(WeakLockFreeCache.java:26)
        at io.opentelemetry.javaagent.bootstrap.field.VirtualFieldImpl$org$apache$spark$scheduler$ActiveJob$io$opentelemetry$javaagent$shaded$io$opentelemetry$context$Context.mapGet(VirtualFieldImplementationsGenerator.java:297)
        at io.opentelemetry.javaagent.bootstrap.field.VirtualFieldImpl$org$apache$spark$scheduler$ActiveJob$io$opentelemetry$javaagent$shaded$io$opentelemetry$context$Context.realGet(VirtualFieldImplementationsGenerator.java)
        at io.opentelemetry.javaagent.bootstrap.field.VirtualFieldImpl$org$apache$spark$scheduler$ActiveJob$io$opentelemetry$javaagent$shaded$io$opentelemetry$context$Context.get(VirtualFieldImplementationsGenerator.java:279)
        at io.opentelemetry.javaagent.instrumentation.spark.ApacheSparkSingletons.getJobContext(ApacheSparkSingletons.java:121)
        at io.opentelemetry.javaagent.instrumentation.spark.v3_1.SparkEventListener.onStageSubmitted(SparkEventListener.java:110)
        at io.opentelemetry.javaagent.instrumentation.spark.v3_1.SparkEventListener.handleSparkListenerEvent(SparkEventListener.java:258)
        at org.apache.spark.scheduler.LiveListenerBus.post(LiveListenerBus.scala:126)
        at org.apache.spark.scheduler.DAGScheduler.submitMissingTasks(DAGScheduler.scala:1547)
        at org.apache.spark.scheduler.DAGScheduler.submitStage(DAGScheduler.scala:1397)
        at org.apache.spark.scheduler.DAGScheduler.handleJobSubmitted(DAGScheduler.scala:1332)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2991)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2982)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2971)
        at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
```

</details>

<details>
<summary>onJobEnd</summary>

```
[otel.javaagent 2024-02-13 19:25:25:507 +0000] [dag-scheduler-event-loop] DEBUG io.opentelemetry.javaagent.bootstrap.ExceptionLogger - Failed to handle exception in instrumentation for org.apache.spark.scheduler.LiveListenerBus
java.lang.NullPointerException
        at io.opentelemetry.javaagent.shaded.instrumentation.api.internal.cache.weaklockfree.AbstractWeakConcurrentMap.getIfPresent(AbstractWeakConcurrentMap.java:113)
        at io.opentelemetry.javaagent.shaded.instrumentation.api.internal.cache.weaklockfree.WeakConcurrentMap.getIfPresent(WeakConcurrentMap.java:48)
        at io.opentelemetry.javaagent.shaded.instrumentation.api.internal.cache.weaklockfree.WeakConcurrentMap$WithInlinedExpunction.getIfPresent (WeakConcurrentMap.java:194)
        at io.opentelemetry.javaagent.shaded.instrumentation.api.internal.cache.WeakLockFreeCache.get(WeakLockFreeCache.java:26)
        at io.opentelemetry.javaagent.bootstrap.field.VirtualFieldImpl$org$apache$spark$scheduler$ActiveJob$io$opentelemetry$javaagent$shaded$io$opentelemetry$context$Context.mapGet(VirtualFieldImplementationsGenerator.java:297)
        at io.opentelemetry.javaagent.bootstrap.field.VirtualFieldImpl$org$apache$spark$scheduler$ActiveJob$io$opentelemetry$javaagent$shaded$io$opentelemetry$context$Context.realGet(VirtualfieldImplementationsGenerator.java)
        at io.opentelemetry.javaagent.bootstrap.field.VirtualFieldImpl$org$apache$spark$scheduler$ActiveJob$io$opentelemetry$javaagent$shaded$io$opentelemetry$context$Context.get(VirtualFieldImplementationsGenerator.java:279)
        at io.opentelemetry.javaagent.instrumentation.spark.ApacheSparkSingletons.getJobContext(ApacheSparkSingletons.java:121)
        at io.opentelemetry.javaagent.instrumentation.spark.v3_1.SparkEventListener.onJobEnd(SparkEventListener.java:81)
        at io.opentelemetry.javaagent.instrumentation.spark.v3_1.SparkEventListener.handleSparkListenerEvent(SparkEventListener.java:256)
        at org.apache.spark.scheduler.LiveListenerBus.post(LiveListenerBus.scala:126)
        at org.apache.spark.scheduler.DAGScheduler.handleTaskCompletion(DAGScheduler.scala:1887)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:3042)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2982)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2971)
        at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
```

</details>

**Comment by keytiong** [comment]
@progxaker. I am not able to reproduce the NullPointerException but I can see how it may occurred. It appears that not all job id has an ActiveJob created. I have refactor the instrumentation logic and hope that change would fix the reported bug.

Are you able to build the extension from the develop branch and test the latest fix? Otherwise, I can publish a SNAPSHOT build.

**Comment by progxaker** [comment]
@keytiong. Now everyting works! I built the jar file from the `develop` branch.
I used Jaeger to show the `spark_job`, `spark_stage` and `spark_task` spans.

<details>
<summary>[For history] Simple Jaeger deployment on K8s</summary>

1. Add the following options to Java:
```
-Dotel.traces.exporter=jaeger
-Dotel.exporter.jaeger.endpoint=http://jaeger:14250
```

2. Run the command `kubectl create -n <namespace> -f <file-with-the-following-context>`.
```yaml
﻿apiVersion: apps/v1
kind: Deployment
metadata:
  name: jaeger-deployment
  labels:
    app: jaeger
spec:
  replicas: 1
  selector:
    matchLabels:
      app: jaeger
  template:
    metadata:
      annotations:
        sidecar.istio.io/inject: "false"
      labels:
        app: jaeger
  spec:
    containers:
    - name: jaeger
      env:
      - name: COLLECTOR_ZIPKIN_HOST_PORT
        value: ":9411"
      image: jaegertracing/all-in-one:1.54
      ports:
      - containerPort: 6831
        name: "zipkin-collect"
        protocol: UDP
      - containerPort: 6832
        name: "zipkin-nodejs"
        protocol: UDP
      - containerPort: 5775
        name: "zipkin-depr"
        protocol: TCP
      - containerPort: 5778
        name: "configs"
        protocol: TCP
      - containerPort: 16686
        name: "frontend"
        protocol: TCP
      - containerPort: 4317
        name: "otlp-http"
        protocol: TCP
      - containerPort: 4318
        name: "otlp-grpc"
        protocol: TCP
      - containerPort: 14268
        name: "jaeger"
        protocol: TCP
      - containerPort: 14250
        name: "model-proto"
        protocol: TCP
      - containerPort: 9411
        name: "zipkin-opt"
        protocol: TCP
---
apiVersion: v1
kind: Service
metadata:
  name: jaeger
  labels:
    app: jaeger
spec:
  ports:
  - port: 6831
    name: "zipkin-collect"
    protocol: UDP
  - port: 6832
    name: "zipkin-nodejs"
    protocol: UDP
  - port: 5775
    name: "zipkin-depr"
    protocol: TCP
  - port: 5778
    name: "configs"
    protocol: TCP
  - port: 16686
    name: "frontend"
    protocol: TCP
  - port: 4317
    name: "otlp-http"
    protocol: TCP
  - port: 4318
    name: "otlp-grpc"
    protocol: TCP
  - port: 14268
    name: "jaeger"
    protocol: TCP
  - port: 14250
    name: "model-proto"
    protocol: TCP
  - port: 9411
    name: "zipkin-opt"
    protocol: TCP
  selector:
    app: jaeger
```

</details>

<details>
<summary>Results</summary>

> The error isn't related to this issue.

![Jaeger Search](https://github.com/keytiong/opentelemetry-spark/assets/26653354/29b6403d-cdb5-487f-afee-6652630c6991)
![Jaeger Trace view](https://github.com/keytiong/opentelemetry-spark/assets/26653354/7971c99c-b7a9-4b1c-ac80-42c7199ac91c)

</details>

I will be working with Microsoft to resolve the problem with Java extensions since they are not loading them.

If you don't mind, we can close this issue since the underlying blocker has been removed.

---

Many thanks!

**Comment by keytiong** [comment]
Hi @progxaker,

Thanks for confirming the fix. I've released version 0.10.0 which includes a few other changes and upgrade to OpenTelemetry 2.1.0 (OTel SDK 1.35.0).

You may want to test that build and raise a new issue if you encountered any other problem.
